server/schedulers/utils.go

```diff
@@ -161,21 +151,26 @@ func takeInfluence(store *core.StoreInfo, storeInfluence *schedule.StoreInfluenc
 }
 
 // shouldBalance returns true if we should balance the source and target store.
-// The min balance diff provides a buffer to make the cluster stable, so that we
+// The tolerantRatio provides a buffer to make the cluster stable, so that we
 // don't need to schedule very frequently.
-func shouldBalance(source, target *core.StoreInfo, kind core.ResourceKind, opInfluence schedule.OpInfluence) bool {
+func shouldBalance(source, target *core.StoreInfo, avgScore float64, kind core.ResourceKind, region *core.RegionInfo, opInfluence schedule.OpInfluence) bool {
 	takeInfluence(source, opInfluence.GetStoreInfluence(source.GetId()))
 	takeInfluence(target, opInfluence.GetStoreInfluence(target.GetId()))
 
-	sourceCount := source.ResourceCount(kind)
 	sourceScore := source.ResourceScore(kind)
 	targetScore := target.ResourceScore(kind)
 	if targetScore >= sourceScore {
 		return false
 	}
-	diffRatio := 1 - targetScore/sourceScore
-	diffCount := diffRatio * float64(sourceCount)
-	return diffCount >= minBalanceDiff(sourceCount)
+
+	// avgScore is the goal for every store
+	// in expection, sourceScore > avgScore > targetScore
+	// if not, moving region is not necessary
+	// In this case, either sourceSizeDiff or targetSizeDiff will be negative, then obviously return false
+	sourceSizeDiff := (sourceScore - avgScore) * source.ResourceWeight(kind)
+	targetSizeDiff := (avgScore - targetScore) * target.ResourceWeight(kind)
+
+	return math.Min(sourceSizeDiff, targetSizeDiff) > float64(region.ApproximateSize)*tolerantRatio
```

Comment:
need we dynamic adjust `tolerantRatio` according to the cluster size like before?